hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/state/StateDefinition.java

```diff
@@ -33,7 +34,77 @@
         int maxKeysHint,
-        boolean onDisk) {}
+        boolean onDisk,
+        boolean singleton) {
+
+    public StateDefinition {
+        if (singleton && onDisk) {
+            throw new IllegalArgumentException("A state cannot both be 'singleton' and 'onDisk'");
+        }
+
+        if (onDisk && maxKeysHint <= 0) {
+            throw new IllegalArgumentException(
+                    "You must specify the maxKeysHint when onDisk. Please see docs.");
+        }
+
+        if (!singleton && keySerdes == null) {
+            throw new NullPointerException(
+                    "keySerdes must be specified when not using singleton types");
+        }
+    }
+
+    /**
+     * Convenience method for creating a {@link StateDefinition} for in-memory k/v states.
+     *
+     * @param stateKey The state key
+     * @param keySerdes The serdes for the key
+     * @param valueSerdes The serdes for the value
+     * @return An instance of {@link StateDefinition}
+     * @param <K> The key type
+     * @param <V> The value type
+     */
+    public static <K extends Comparable<K>, V> StateDefinition<K, V> inMemory(
+            @NonNull final String stateKey,
+            @NonNull final Serdes<K> keySerdes,
+            @NonNull final Serdes<V> valueSerdes) {
+        return new StateDefinition<>(stateKey, keySerdes, valueSerdes, -1, false, false);
+    }
+
+    /**
+     * Convenience method for creating a {@link StateDefinition} for on-disk k/v states.
+     *
+     * @param stateKey The state key
+     * @param keySerdes The serdes for the key
+     * @param valueSerdes The serdes for the value
+     * @param maxKeysHint A hint as to the maximum number of keys to be stored in this state. This
+     *     value * CANNOT CHANGE from one schema version to another. If it is changed, you will need
+     *     to do a * long-form migration to a new state.
+     * @return An instance of {@link StateDefinition}
+     * @param <K> The key type
+     * @param <V> The value type
+     */
+    public static <K extends Comparable<K>, V> StateDefinition<K, V> onDisk(
+            @NonNull final String stateKey,
+            @NonNull final Serdes<K> keySerdes,
+            @NonNull final Serdes<V> valueSerdes,
+            final int maxKeysHint) {
+        return new StateDefinition<>(stateKey, keySerdes, valueSerdes, maxKeysHint, true, false);
+    }
+
+    /**
+     * Convenience method for creating a {@link StateDefinition} for singleton states.
+     *
+     * @param stateKey The state key
+     * @param valueSerdes The serdes for the singleton value
+     * @return An instance of {@link StateDefinition}
+     * @param <K> The key type
+     * @param <V> The value type
+     */
+    public static <K extends Comparable<K>, V> StateDefinition<K, V> singleton(
+            @NonNull final String stateKey, @NonNull final Serdes<V> valueSerdes) {
+        return new StateDefinition<>(stateKey, null, valueSerdes, 1, false, true);
```

Comment:
Sorry if this is obvious. But why is `maxKeysHint` 1 here ? Shouldn't it be set only for `onDisk` singletons ?

Reply:
No reason at all, TBH. The field is ignored, but if it *were* to have a meaning, I thought "1" was true. But I don't think it matters. Maybe -1 would be better so nobody thinks there is a meaning to it.

Reply:
Changed to -1 (with a static constant `NO_MAX`)